tests/assembly-llvm/stack-protector/stack-protector-heuristics-effect-2.rs

```diff
@@ -0,0 +1,481 @@
+//@ revisions: all strong none missing
+//@ assembly-output: emit-asm
+//@ ignore-apple slightly different policy on stack protection of arrays
+//@ ignore-msvc stack check code uses different function names
+//@ ignore-nvptx64 stack protector is not supported
+//@ ignore-wasm32-bare
+//@ [all] compile-flags: -Z stack-protector=all
+//@ [strong] compile-flags: -Z stack-protector=strong
+//@ [none] compile-flags: -Z stack-protector=none
+//@ compile-flags: -C opt-level=2 -Z merge-functions=disabled
+
+#![crate_type = "lib"]
+#![allow(internal_features)]
+#![feature(unsized_fn_params)]
+
+extern "C" {
+    fn strcpy(dest: *mut u8, src: *const u8) -> *mut u8;
+    fn printf(fmt: *const u8, ...) -> i32;
+    fn funcall(p: *mut i32);
+    fn funcall2(p: *mut *mut i32);
+    fn funfloat(p: *mut f64);
+    fn funfloat2(p: *mut *mut f64);
+    fn testi_aux() -> f64;
+    fn getp() -> *mut i32;
+    fn dummy(_: ...) -> i32;
+
+    static STR: [u8; 1];
+}
+
+extern "C-unwind" {
+    fn except(p: *mut i32);
+}
+
+#[repr(C)]
+struct Pair {
+    a: i32,
+    b: i32,
+}
+
+#[repr(C)]
+struct Nest {
+    first: Pair,
+    second: Pair,
+}
+
+// test1: array of [16 x i8]
+// CHECK-LABEL: test1{{:|\[}}
+#[no_mangle]
+pub fn test1(a: *const u8) {
+    let mut buf: [u8; 16] = [0; 16];
+
+    unsafe {
+        strcpy(buf.as_mut_ptr(), a);
+        printf(STR.as_ptr(), buf.as_ptr());
+    }
+
+    // all: __stack_chk_fail
+    // strong: __stack_chk_fail
+    // none-NOT: __stack_chk_fail
+    // missing-NOT: __stack_chk_fail
+}
+
+// test3: array [4 x i8]
+// CHECK-LABEL: test3{{:|\[}}
+#[no_mangle]
+pub fn test3(a: *const u8) {
```

Comment:
Since we delete some unneccesary tests, I think these function names are no need to be consistent with ones in llvm's tests. You can just name `test2` or `test_array4_i8`.

Same for all  functiosn below